### ISSUE TITLE
L2SCA

### ISSUE THREAD

**Issue by elefantinerrrrs** [shivam5992/textstat#173]
Hi, I am interested in implementing an L2 Syntactic Complexity Analyzer for _textstat_. It automates the analysis of texts written by second language learners, which helps both researchers of L2 learners and L2 teachers in the classroom by providing 14 indices describing the syntactic complexity of the text. 
The following calculations are necessary to implement the L2SCA:
- number of words
- ! number of clauses
- ! number of dependant clauses
- ! number of T-Units
- ! number of complex T-Units
- number of sentences
- ! number of coordinate phrases
- ! number of verb phrases
- ! number of complex nominals

In addition to the above, (Lu, 2010) uses the following tools to count them:
- ! A syntactic parser from which they derive syntax structure trees of each sentence in the text. Their choice was Stanford Parser
- ! Tregex to extract patterns from the trees

A ! denotes something that _textstat_ does not currently offer, which would entail additional work in the process. 

A web based version is hosted here: https://aihaiyang.com/software/l2sca/

Link to papers and the original, local implementation: http://www.personal.psu.edu/xxl13/downloads/l2sca.html 
https://www.jbe-platform.com/docserver/fulltext/ijcl.15.4.02lu.pdf 

Is this something we would like to add to _textstat_?

**Comment by Clickedbigfoot** [comment]
Indeed. I do not know if that is an acceptable expectation of a user of _textstat_, but I'm thinking of using [spacy](https://spacy.io/usage/linguistic-features#dependency-parse) instead. This still requires a download of a pretrained model, but I don't believe it would be acting as a server (And therefore wouldn't need any interaction outside of _textstat_ and the initial download): 
`import spacy
nlp = spacy.load("en_core_web_sm")
doc = nlp("I put the book in the box on the table")`

The download would be
`python -m spacy download en_core_web_sm`
It's worth noting that this doesn't seem to have the exact same results as Stanford though. In the example sentence above, Stanford has "put" as a child of "I", but spacy would have "I" as a child of "put", so that means that Stanford treats NP and VP as children of S, but spacy treats NP as children of VP

**Comment by LKirst** [comment]
Would it be feasible to give the user a choice of back end parsers? Then, experienced users could try to replicate the Lu (2010) results and refer to the paper to explain the measure they used.

**Comment by Clickedbigfoot** [comment]
A choice would be nice. Ideally, both parsers would replicate the same scores as Lu (2010), but they would just pose different setup for the user and different implementation of how each parser is used within textstat (Lu has formal definitions for each counted syntactic structure that help count them from the parsed tree. I'm looking into if different, but similarly reliable formal definitions can be used to count them from the dependency tree returned by spacy)

**Comment by Clickedbigfoot** [comment]
Without having looked into this as much as I would have liked to by now, I realize that spacy also includes [spacy.matcher](https://spacy.io/usage/rule-based-matching). This could solve the task of counting the meaningful units from the spacy parser without importing a separate package (Ex: Tregex) following an example [here for extracting verb phrases](https://stackoverflow.com/questions/47856247/extract-verb-phrases-using-spacy). This could also help for counting prepositional phrases in #132

**Comment by BrandonKMLee** [comment]
Sounds good (thx for the tag BTW)
